File: cobalt/backend/presentation/contracts/http/routers/servers.py
#  Copyright (C) 2026 ArtoriasCode
#  Author: ArtoriasCode
#  Repository: https://github.com/ArtoriasCode/cobalt
#  SPDX-License-Identifier: AGPL-3.0-or-later

from abc import abstractmethod, ABC
from typing import Any

from presentation.contracts.http.routers import AbstractHttpRouter


class AbstractHttpServersRouter(AbstractHttpRouter, ABC):
    """
    Abstract HTTP router for servers-related operations.
    """

    @abstractmethod
    async def get_page(self, *args: Any, **kwargs: Any) -> Any:
        """
        Gets a paginated list of servers.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def get_one(self, *args: Any, **kwargs: Any) -> Any:
        """
        Gets an existing server.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def create_one(self, *args: Any, **kwargs: Any) -> Any:
        """
        Creates a new server.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def update_one(self, *args: Any, **kwargs: Any) -> Any:
        """
        Updates an existing server.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def delete_one(self, *args: Any, **kwargs: Any) -> Any:
        """
        Deletes an existing server.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def delete_many(self, *args: Any, **kwargs: Any) -> Any:
        """
        Deletes multiple existing servers.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def start(self, *args: Any, **kwargs: Any) -> Any:
        """
        Starts an existing server.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def stop(self, *args: Any, **kwargs: Any) -> Any:
        """
        Stops an existing server.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def execute(self, *args: Any, **kwargs: Any) -> Any:
        """
        Executes a command inside the server container.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

    @abstractmethod
    async def status(self, *args: Any, **kwargs: Any) -> Any:
        """
        Gets the server container status.

        Parameters:
        - *args: Positional arguments.
        - **kwargs: Keyword arguments.

        Returns:
        - Any: Response object.
        """
        ...

File: cobalt/frontend/src/services/api/http/servers.ts
/*
 * Copyright (C) 2026 ArtoriasCode
 * Author: ArtoriasCode
 * Repository: https://github.com/ArtoriasCode/cobalt
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import type { IHttpClient, IHttpServersApiService } from "@/contracts"
import type {
  ServerEntity,
  ServersPageEntity,
  ServerStatusEntity,
  ServersPageRequest,
  ServerCreateRequest,
  ServerUpdateRequest,
  ServerExecuteRequest
} from "@/types"

/**
 * API service for servers endpoints.
 */
export class HttpServersApiService implements IHttpServersApiService {
  private readonly prefix = "/api/v1/servers"
  private readonly client: IHttpClient

  constructor(client: IHttpClient) {
    this.client = client
  }

  /**
   * Gets a paginated list of servers.
   *
   * Parameters:
   * - params: ServersPageRequest object.
   *
   * Returns:
   * - Promise<ServersPageResponse>: Paginated list of servers.
   */
  async getPage(params: ServersPageRequest): Promise<ServersPageEntity> {
    return this.client.get<ServersPageEntity>(`${this.prefix}/`, { params })
  }

  /**
   * Gets an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<ServerEntity>: Server entity.
   */
  async getOne(serverId: number): Promise<ServerEntity> {
    return this.client.get<ServerEntity>(`${this.prefix}/${serverId}`)
  }

  /**
   * Creates a new server.
   *
   * Parameters:
   * - data: ServerCreateRequest object.
   *
   * Returns:
   * - Promise<ServerEntity>: Created server entity.
   */
  async createOne(data: ServerCreateRequest): Promise<ServerEntity> {
    return this.client.post<ServerEntity>(`${this.prefix}/`, data)
  }

  /**
   * Updates an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   * - data: ServerUpdateRequest object.
   *
   * Returns:
   * - Promise<ServerEntity>: Updated server entity.
   */
  async updateOne(serverId: number, data: ServerUpdateRequest): Promise<ServerEntity> {
    return this.client.patch<ServerEntity>(`${this.prefix}/${serverId}`, data)
  }

  /**
   * Deletes an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  async deleteOne(serverId: number): Promise<void> {
    return this.client.delete<void>(`${this.prefix}/${serverId}`)
  }

  /**
   * Deletes multiple existing servers.
   *
   * Parameters:
   * - ids: List of server IDs.
   *
   * Returns:
   * - Promise<void>.
   */
  async deleteMany(ids: number[]): Promise<void> {
    return this.client.delete<void>(`${this.prefix}/batch`, { data: ids })
  }

  /**
   * Starts an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  async start(serverId: number): Promise<void> {
    return this.client.post<void>(`${this.prefix}/${serverId}/start`)
  }

  /**
   * Stops an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  async stop(serverId: number): Promise<void> {
    return this.client.post<void>(`${this.prefix}/${serverId}/stop`)
  }

  /**
   * Restarts an existing server.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<void>.
   */
  async restart(serverId: number): Promise<void> {
    return this.client.post<void>(`${this.prefix}/${serverId}/restart`)
  }

  /**
   * Executes a command inside the server container.
   *
   * Parameters:
   * - serverId: Server ID.
   * - data: ServerExecuteRequest object.
   *
   * Returns:
   * - Promise<void>.
   */
  async execute(serverId: number, data: ServerExecuteRequest): Promise<void> {
    return this.client.post<void>(`${this.prefix}/${serverId}/execute`, data)
  }

  /**
   * Gets the server container status.
   *
   * Parameters:
   * - serverId: Server ID.
   *
   * Returns:
   * - Promise<ServerStatusEntity>: Server status entity.
   */
  async status(serverId: number): Promise<ServerStatusEntity> {
    return this.client.get<ServerStatusEntity>(`${this.prefix}/${serverId}/status`)
  }
}